### PR TITLE
sdi-494  initial commit of service that subscribes and logs all domai…

### PR DESCRIPTION
**Pull request by lbennett-moj**
…n events to App Insights as custom events.

### CODE REVIEW

File: helm_deploy/values-dev.yaml [hunk 1]
```diff
@@ -9,8 +9,13 @@ generic-service:
 
   env:
     APPLICATIONINSIGHTS_CONFIGURATION_FILE: applicationinsights.dev.json
+    API_BASE_URL_OAUTH: https://sign-in-dev.hmpps.service.justice.gov.uk/auth
```
Comment (lbennett-moj) [inline helm_deploy/values-dev.yaml:12]: requires auth for the HMPPS SQS library (providing queue maintenance endpoints)

File: src/main/kotlin/uk/gov/justice/digital/hmpps/domaineventlogger/DomainEventsListener.kt [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+  /* Domain events have top level attributes and some optional nested objects - eg AdditionalInformation. */
+  fun translateMap(rawMap: Map<String, Any>): Map<String, String> {
+    val mapNested = rawMap.filter { it.value is Map<*, *> } as Map<String, Map<*, *>>
+    // for nested properties use top level key and nested key seperated by a full stop
+    val flattenedNestedEntries: List<Map.Entry<String, String>> =
+      mapNested.map { (key, value) -> value.entries.associate { "$key.${it.key}" to it.value.toString() } }
+        .flatMap { it.entries }
+    return rawMap.filter { it.value !is Map<*, *> }.mapValues { it.value.toString() }.toMutableMap()
+      .plus(flattenedNestedEntries.map { it.toPair() }.toMap())
+  }
```
Comment (lbennett-moj) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/domaineventlogger/DomainEventsListener.kt:45]: will come back to this nastiness to simplify, once deployment checked

File: docker-compose-localstack.yaml [hunk 3]
```diff
@@ -0,0 +1,20 @@
+version: "3"
+services:
+  localstack:
+    image: localstack/localstack:0.14.2
```
Comment (mikehalmamoj) [inline docker-compose-localstack.yaml:4]: Might as well start out with the latest version?

File: docker-compose-localstack.yaml [hunk 4]
```diff
@@ -0,0 +1,20 @@
...
+    networks:
+      - hmpps
+    container_name: localstack
+    ports:
+      - "4566-4597:4566-4597"
```
Comment (mikehalmamoj) [inline docker-compose-localstack.yaml:9]: I think you only need 4566 now.

File: src/test/kotlin/uk/gov/justice/digital/hmpps/domaineventlogger/DomainEventsListenerTest.kt [hunk 5]
```diff
@@ -0,0 +1,98 @@
...
+      },
+      isNull()
+    )
+  }
+
+  @Test
+  internal fun `can handle unexpected data`() {
+    listener.onDomainEventReceived(
+      message = invalidDomainEventMessage()
+    )
```
Comment (mikehalmamoj) [inline src/test/kotlin/uk/gov/justice/digital/hmpps/domaineventlogger/DomainEventsListenerTest.kt:49]: I really like the fact we are testing bad data - but could we add some tests for very very bad data? I'm thinking that even if we get malformed JSON we should create telemetry for it including its contents.

**Review by mikehalmamoj — APPROVED** [review]

**Pull request merged**